[2단계 - 로또 리팩토링] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
안녕하세요 제이!
피드백 주신 부분 반영하여 리팩토링하였습니다.

### 리팩토링한 부분
- 변수 인라인으로 수정
- 접근제어자 변경
- 메서드 위치 수정
- 부족한 테스트 추가
- 버그 수정

### 궁금한 점
1. `Lotto` 생성 책임을 `Lotto`가 아닌 `RandomLottoGenerator` 가 가지는 것이 좋은 설계인지 궁금합니다.
```java
public class RandomLottoGenerator {

    public List<Lotto> generate(final int count) {
        return IntStream.range(0, count)
                .mapToObj(number -> makeRandomNumbers())
                .map(HashSet::new)
                .map(Lotto::new)
                .toList();
    }
```
현재 `Lotto` 객체를 생성하는 방법에는 1. `Lotto` 생성자 이용하기 2. `RandomLottoGenerator`의 `generate()` 이용하기 2가지 방법이 있습니다. 전자는 로또를 수동으로 만들 때(우승 로또 번호), 후자는 로또를 랜덤으로 생성할 때 사용합니다.

그런데 `RandomLottoGenerator`가 굳이 필요할까? 라는 생각이 들었습니다. **`Lotto` 객체가 자기 자신을 생성하는 책임을 가져야되는 것이 아닌가?** 라는 고민을 하게 되었는데요. 디자인 패턴을 살펴보니 `팩토리 메서드`, `추상 팩토리` 패턴에서는 객체 생성 책임을 분리해서 확장성을 높이고 유연하게 설계하고 있는 것을 발견했습니다. 

제이가 생각하기에 이 구조가 적합한지(**객체 생성 책임을 다른 객체가 가지고 있는 것이 좋은 설계라고 생각하시는지**) 궁금합니다.
> 구조가 적절하지 않다면 클래스 내에 정적 팩토리 생성자를 이용해서 생성하는 방법으로 변경하려고 합니다.
   정적 팩토리 생성자를 이용해보았더니 생성하는 부분이 너무 복잡해지는 단점이 있었습니다.

2. **설계 공부를 하려고하는데, 추천해주실 책이나 웹사이트 있다면 공유 부탁드립니다!**


감사합니다! 🙇‍♀️

Starlight258 committed:
refactor(LottoController): 변수 인라인으로 수정

Starlight258 committed:
refactor(LottoController): 상수 접근제어자 변경

- 클래스 내부에서만 사용되는 변수는 접근제어자를 private으로 설정한다.

Starlight258 committed:
refactor(WinningResult): getter 위치 수정 및 방어적 복사 수행

- getter를 java 권장 메서드 순서에 따라 맨 아래에 둔다.
- 외부에서 내부 맵을 변경하는 것을 막기 위해 방어적 복사를 수행하여 반환한다.

Starlight258 committed:
refactor(WinningResultCalculator): 호출되는 순서에 따라 메서드 순서 변경

Starlight258 committed:
refactor(All): 메서드, 변수명 수정 및 중복 메서드 제거

Starlight258 committed:
refactor(LottoController): 메서드 추출

Starlight258 committed:
refactor(All): 변수에 final 키워드 추가

- 코드의 예측 가능성을 높이고 스레드 안전, 성능 상의 이점으로 인해 final 키워드를 붙인다.

Starlight258 committed:
refactor(Lotto): 메서드 추출

Starlight258 committed:
fix(OutputView): 당첨 통계 출력시 2등의 경우 보너스 볼 일치 메세지도 출력하도록 수정

Starlight258 committed:
refactor(RandomLottoGenerator): 클래스명 수정 및 메서드 수정

- 로또 번호를 랜덤으로 추출하여 로또를 생성하기 때문에 클래스명에 Random을 추가한다.

Starlight258 committed:
test(LottoNumber): 테스트 케이스 추가 및 누락된 final 키워드 추가

Starlight258 committed:
refactor(Lotto): 메서드, 상수 위치 변경

- 메서드를 public -> protected -> default -> private 순으로 정렬한다. (equals, getter/setter는 뒤에 위치)
- 로또 크기는 로또 데이터와 관련된 상수이므로 Lotto 클래스로 위치를 옮긴다.

Starlight258 committed:
feat(LottoNumber): 캐싱 도입하여 객체 재사용

- LottoNumber는 불변 객체이고, 1~45 범위의 숫자가 매번 생성되므로 캐싱을 도입하여 메모리를 절약한다.

Starlight258 committed:
refactor(RandomLottoGenerator): Stream 사용하도록 변경

- Stream을 이용하여 가독성을 향상한다.

Starlight258 committed:
refactor(LottoPrice): 빠진 테스트 추가

Starlight258 committed:
test(Lotto): 빠진 테스트 추가

Starlight258 committed:
refactor(WinningResult): 메서드명 명확하게 수정

Starlight258 committed:
refactor(LottoController): 메서드명 변경

Starlight258 committed:
refactor(InputValidator): InputValidator 클래스 view 패키지로 이동

- InputValidator는 InputView에서만 사용되므로 util 패키지에 둘만큼 범용적이지 않다.

Starlight258 committed:
test(InputValidator, StringParser): 정상 케이스 테스트 추가

Starlight258 committed:
test(LottoNumber, LottoPrice): 테스트 메서드명 변경

Starlight258 committed:
test(LottoPrice, StringParser): given절에 테스트에 필요한 데이터 명시하도록 수정

Starlight258 committed:
test(LottoPrice): 검증 정상적으로 수행되도록 수정

Starlight258 committed:
test(WinningResultCalculator): given절에 테스트 기대 결과도 포함하도록 수정

- given절에는 테스트에 필요한 모든 데이터(전제 조건, 기대 결과 등등)를 포함하도록 한다.

Starlight258 committed:
refactor(Lotto): 내부 필드 자료구조를 List에서 Set으로 변경

- 로또 번호는 중복되지 않은 6개의 숫자여야하므로 Set을 이용하면 중복 여부를 요소의 개수로 간편하게 확인할 수 있다.

Starlight258 committed:
test(StringParser): 필요없는 테스트 설명 제거

JunHoPark93 commented on src/main/java/lotto/domain/RandomLottoGenerator.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class RandomLottoGenerator {
+
+    private static final int START_INDEX = 0;
+
+    public List<Lotto> generate(final int count) {
+        return IntStream.range(0, count)
+                .mapToObj(number -> makeRandomNumbers())
+                .map(HashSet::new)
+                .map(Lotto::new)
+                .toList();
+    }
```
> 그런데 RandomLottoGenerator가 굳이 필요할까? 라는 생각이 들었습니다. Lotto 객체가 자기 자신을 생성하는 책임을 가져야되는 것이 아닌가? 라는 고민을 하게 되었는데요. 디자인 패턴을 살펴보니 팩토리 메서드, 추상 팩토리 패턴에서는 객체 생성 책임을 분리해서 확장성을 높이고 유연하게 설계하고 있는 것을 발견했습니다.
> 
> 제이가 생각하기에 이 구조가 적합한지(객체 생성 책임을 다른 객체가 가지고 있는 것이 좋은 설계라고 생각하시는지) 궁금합니다.

정답은 없지만 로또를 생성하는 책임을 다른 객체에 부여하는게 어색하진 않다고 생각합니다. 생성기라는것을 정의해도 되니까요. 다만 현재 구조에서 테스트를 어떻게 더 보강하면 좋을지, 인스턴스화가 필요한것일지 같이 고민해보면 어떨까요?

Starlight258 replied:
로또 생성기의 예외 테스트가 없었네요! 추가하겠습니다.
`RandomLottoGenerator`를 `LottoGenerator` 인터페이스를 구현하도록 하면 로또 생성 전략이 달라지더라도 변경점이 줄어들겠네요! 수정하겠습니다!

JunHoPark93 commented on src/main/java/lotto/domain/LottoPrice.java:
```diff
@@ -15,7 +15,7 @@ public int calculateLottoCount() {
     }
 
-    public double divideFrom(int inputAmount) {
+    public double divideFrom(final int inputAmount) {
```
LottoPrice 클래스내부에 amount라는 인스턴스변수와 UNIT_PRICE_OF_LOTTO 를 가지고있는 로또금액이라는 도메인을 만들어주신걸로 이해했습니다. 로또의 금액을 표현하려는거같은데 로또의 개수를 구하거나 나눗셈의 역할을 부여하셨는데요, 역할자체가 불분명한것처럼 느껴지는데 어떻게 생각하시나요?

Starlight258 replied:
확실히 `LottoPrice`는 **가격 범위 검증**과 **나눗셈**의 역할만 수행해서 모호하네요! 가격으로 로또 개수를 구하는 것도 `LottoPrice`의 책임이라고 보기보다는 `LottoGenerator`(로또 생성기)의 책임에 더 가까운 것 같아요. 

`LottoPrice` 객체를 만들기보다, `RandomLottoGenerator`에서 가격을 받아 로또를 생성하는 방식으로 구현하는 게 좋아보여요!

JunHoPark93 commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -1,25 +1,22 @@
-
     public double calculateProfitRate(final LottoPrice lottoPrice) {
-        return lottoPrice.divideFrom(calculateTotalAmount());
+        return lottoPrice.divideFrom(calculateTotalProfit());
```
calculateProfitRate 메서드를 보면 LottoPrice를 내부로 들고들어가서 해당 객체의 메서드를 호출하고있는데요, 클래스간 불필요한 의존성이 생기는 모습으로 보입니다.
LottoPrice의 역할을 명확히정의하고 수익률 계산을 책임지게 한다면 두 클래스의 의존성을 분리할 수 있을것으로 보이는데요, 

LottoPrice내부에 calculateProfitRate(final int totalProfit) {...} 등을 정의해서 해결해보면 어떨까요?

Starlight258 replied:
확인해보니 `WinningResult`에서 단순히 `LottoPrice`의 메서드를 호출한다면 불필요한 의존성이 생기게 되네요 !!

현재 코드는 위에서 언급하신대로 `LottoPrice`가 모호한 책임을 갖는다고 판단되어 지운 상태여서 해당 문제는 자연스럽게 해결된 것 같아요. 감사합니다!

JunHoPark93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -38,19 +22,20 @@ public int countMatchingLottoNumber(final Lotto lotto) {
     }
 
-    private List<LottoNumber> makeNumber(final List<Integer> values) {
+    private void validateSize(final Set<Integer> values) {
```
해당 클래스 내의 접근제한자가 default인 메서드들은 의도하신거죠?

Starlight258 replied:
네! 같은 패키지인 로또 도메인 내에서만 사용하는 상수이기 때문에 `public` 대신 `default` 접근제어자를 사용했습니다 !

JunHoPark93 commented on src/main/java/lotto/domain/LottoAward.java:
```diff
@@ -18,25 +18,25 @@ public enum LottoAward {
+    private final boolean matchesBonusNumber;
 
-    LottoAward(final int matchingCount, final int amount, final boolean isBonusNumberMatched) {
+    LottoAward(final int matchingCount, final int amount, final boolean matchesBonusNumber) {
```
정의해주신 도메인들에 대해서는 모두 테스트코드를 작성해주세요~

Starlight258 replied:
`enum`도 테스트를 해야하는군요! 작성하겠습니다!

JunHoPark93 commented:
> 설계 공부를 하려고하는데, 추천해주실 책이나 웹사이트 있다면 공유 부탁드립니다!

웹상에 참고할만한 양질의 자료가 많은거같은데, 저는 클린코드 라는 책이 도움이 되었었습니다.

JunHoPark93 reviewed (changes requested):
안녕하세요 밍트~
2단계 구현하느라 고생많으셨어요.
몇가지 코멘트 남겼는데 확인 후 의견 및 반영부탁드릴게요~

Starlight258 committed:
fix(RandomLottoGenerator): 로또 생성 개수가 양수가 아닌 경우 예외 발생하도록 수정

Starlight258 committed:
feat(LottoGenerator): LottoGenerator 인터페이스를 의존하도록 하여 DIP 지키도록 구현

Starlight258 committed:
refactor(LottoPrice): 모호한 책임을 가진 LottoPrice 삭제

- LottoPrice에서 가격 검증 및 로또 개수 계산 로직을 LottoGenerator가 갖도록 수정한다.

Starlight258 committed:
fix(LottoAward): 보너스볼이 일치하는 경우 상금이 제대로 계산되지 않은 버그 수정

- 보너스 볼이 일치하고 6개 일치한 경우 1등으로 판단되지 않는데, 이는 보너스 볼 판단 여부를 나타내는 matchesBonusNumber가 기본값으로 false로 설정되어 있기 때문이다.
- 따라서 matchesBonusNumber 필드를 없애고 적절히 상금이 계산되도록 수정한다.

Starlight258 committed:
test(LottoAward): 테스트 작성

Starlight258 commented:
@JunHoPark93 리뷰 반영했습니다! 감사드립니다!

JunHoPark93 commented on src/test/java/lotto/domain/LottoAwardTest.java:
```diff
@@ -0,0 +1,27 @@
+
+class LottoAwardTest {
+
+    @ParameterizedTest
```
👍

JunHoPark93 reviewed (approved):
안녕하세요 밍트~
리뷰 반영하느라 고생많으셨습니다.
이만 머지할게요 다음 미션도 화이팅입니다!

Starlight258 commented:
@JunHoPark93 감사합니다 많이 배웠습니다!  🙇‍♀️
